### PR TITLE
Update TF Decision Forests example

### PR DESCRIPTION
**Pull request by ksalama**
Hello @fchollet and @achoum - 

Please review the updates I made to the TF Decision Forests example.

Cheers

### CODE REVIEW

File: examples/structured_data/classification_with_tfdf.py [hunk 1]
```diff
@@ -126,14 +115,39 @@
...
+
 """
 ## Configure hyperparameters
 
 You can find all the parameters of the Gradient Boosted Tree model in the
 [documentation](https://www.tensorflow.org/decision_forests/api_docs/python/tfdf/keras/GradientBoostedTreesModel)
 """
 
-GROWING_STRATEGY = "BEST_FIRST_GLOBAL"
 NUM_TREES = 250
```
Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:151]: Please remove.
See comment in https://github.com/keras-team/keras-io/pull/782

File: examples/structured_data/classification_with_tfdf.py [hunk 2]
```diff
@@ -100,6 +87,8 @@
 
 # Target column name.
 TARGET_COLUMN_NAME = "income_level"
+# The labels of the target columns.
+TARGET_LABELS = [" - 50000.", " 50000+."]
 # Weight column name.
 WEIGHT_COLUMN_NAME = "instance_weight"
```
Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:93]: Please remove.
See comment in https://github.com/keras-team/keras-io/pull/782

File: examples/structured_data/classification_with_tfdf.py [hunk 3]
```diff
@@ -146,23 +160,14 @@
...
-def prepare_sample(features, target, weight):
-    for feature_name in features:
-        if feature_name in CATEGORICAL_FEATURES_WITH_VOCABULARY:
-            if features[feature_name].dtype != tf.dtypes.string:
-                # Convert categorical feature values to string.
-                features[feature_name] = tf.strings.as_string(features[feature_name])
-    return features, target, weight
-
-
 def run_experiment(model, train_data, test_data, num_epochs=1, batch_size=None):
```
Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:163]: What about implementing the "Nit: What about having a **fit_args argument to the run_experiment function?" discussed before.

**Comment by fchollet** [comment]
@achoum @ksalama are we good here?

File: examples/structured_data/classification_with_tfdf.py (hunk 2 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:93]: The `TARGET_LABELS` is used later in the `prepare_dataframe` method.

File: examples/structured_data/classification_with_tfdf.py (hunk 3 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:163]: Most of the Keras examples use the current `run_experiment` signature, and I didn't want to deviate from them.
Are you OK with that?

File: examples/structured_data/classification_with_tfdf.py (hunk 1 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:151]: I just wanted to expose some of the key parameters of the algorithms to be visible to the readers. I can add descriptions as comments to each one. Also happy to remove them. What do you think?

File: examples/structured_data/classification_with_tfdf.py [hunk 4]
```diff
@@ -146,23 +160,14 @@
...
         train_data, label=TARGET_COLUMN_NAME, weight=WEIGHT_COLUMN_NAME
-    ).map(prepare_sample, num_parallel_calls=tf.data.AUTOTUNE)
+    )
     test_dataset = tfdf.keras.pd_dataframe_to_tf_dataset(
         test_data, label=TARGET_COLUMN_NAME, weight=WEIGHT_COLUMN_NAME
-    ).map(prepare_sample, num_parallel_calls=tf.data.AUTOTUNE)
+    )
 
     model.fit(train_dataset, epochs=num_epochs, batch_size=batch_size)
     _, accuracy = model.evaluate(test_dataset, verbose=0)
```
Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:173]: At line 222 (for some reason, I cannot create a comment there):

You are effectively replicating the detection of the semantic of the features. This version is relatively simple and will not handle many cases. It is better to let TF-DF do the semantic detection.

Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:173]: **Cannot comment in this block anymore.

>> Sort

I think tf.sort and unique are done in memory.
It would be interesting to understand and document what append in this processing layer when used on a large dataset. For example, will the entire dataset be loaded in memory?

File: examples/structured_data/classification_with_tfdf.py [hunk 5]
```diff
@@ -146,23 +160,14 @@
...
     train_dataset = tfdf.keras.pd_dataframe_to_tf_dataset(
         train_data, label=TARGET_COLUMN_NAME, weight=WEIGHT_COLUMN_NAME
-    ).map(prepare_sample, num_parallel_calls=tf.data.AUTOTUNE)
+    )
     test_dataset = tfdf.keras.pd_dataframe_to_tf_dataset(
         test_data, label=TARGET_COLUMN_NAME, weight=WEIGHT_COLUMN_NAME
-    ).map(prepare_sample, num_parallel_calls=tf.data.AUTOTUNE)
+    )
 
     model.fit(train_dataset, epochs=num_epochs, batch_size=batch_size)
```
Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:172]: This division will be unstable for rare categorical values. With target mean, it works well to add a correction term controlled by an hyperparameter.

correction =  1
positive_probability = positive_frequency / (positive_frequency + negative_frequency + correction)

Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:172]: In this example, I would do the division in double (float64) and then cast the result to float32.

File: examples/structured_data/classification_with_tfdf.py [hunk 6]
```diff
@@ -383,22 +387,22 @@ def call(self, inputs):
...
-        positive_probability = positive_fequency / (
-            positive_fequency + negative_fequency
+        positive_probability = positive_frequency / (
+            positive_frequency + negative_frequency
         )
         # Concatenate and return the looked-up statistics.
         return tf.concat(
-            [positive_fequency, negative_fequency, positive_probability], axis=1
+            [positive_frequency, negative_frequency, positive_probability], axis=1
         )
```
Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:407]: Copy from https://github.com/keras-team/keras-io/pull/782

Could the colab report the test accuracy of the three tested methods?

Example: https://www.tensorflow.org/decision_forests/tutorials/model_composition_colab#whats_next

I don't see the output of the colab in this review.


>> TensorFlow Decision Forests provide powerful models, especially with structured data.
In our experiments, the Gradient Boosted Tree model achieved 95.79% test accuracy.
When using the target encoding with categorical feature, the same model achieved 95.81% test accuracy.
When pretraining embeddings to be used as inputs to the Gradient Boosted Tree model,
we achieved 95.82% test accuracy.

If the original model has a better accuracy than the model with pre-processing, this might indicate a bug in the code. I would not leave it this way. Could it be the smoothing hyper-parameter?

Alternatively, I would separate the simple GBT code from the experimental pre-processing in two separate examples (files). Users that just want to learn about GBT would benefit from it.

File: examples/structured_data/classification_with_tfdf.py [hunk 7]
```diff
@@ -383,22 +387,22 @@ def call(self, inputs):
...
         # Compute positive probability for the input feature values.
-        positive_probability = positive_fequency / (
-            positive_fequency + negative_fequency
+        positive_probability = positive_frequency / (
+            positive_frequency + negative_frequency
         )
         # Concatenate and return the looked-up statistics.
         return tf.concat(
-            [positive_fequency, negative_fequency, positive_probability], axis=1
+            [positive_frequency, negative_frequency, positive_probability], axis=1
```
Comment (achoum) [inline examples/structured_data/classification_with_tfdf.py:405]: Can you address the comment starting with:

> This is not really an embedding. Instead, it will be the output of the linear model trained with categorical buckets. So it is fundamentally very similar to target mean.

in https://github.com/keras-team/keras-io/pull/782

File: examples/structured_data/classification_with_tfdf.py (hunk 1 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:151]: Do you think it is better to use the [documentation](https://www.tensorflow.org/decision_forests/api_docs/python/tfdf/keras/GradientBoostedTreesModel) link instead? (next to the model construction object)

File: examples/structured_data/classification_with_tfdf.py (hunk 2 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:93]: I am not using the attribute "as an input feature", rather as a weight to each instance. This will affect the loss computation; instances with higher weight will contribute more to the loss. This enforces the model to try to get the highly-weighted instances right, more than the low-weight instances, and I think it should be used as it represents the true distribution of the data.

Happy to discuss further

File: examples/structured_data/classification_with_tfdf.py (hunk 5 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:172]: Done

Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:172]: Will do

File: examples/structured_data/classification_with_tfdf.py (hunk 1 above)
Comment (achoum) [reply examples/structured_data/classification_with_tfdf.py:151]: Yes it is equally good.

Note that this documentation pages also list all the Keras model arguments (not just the hyper-parameters) and it does not list the possible values the say way.

File: examples/structured_data/classification_with_tfdf.py (hunk 2 above)
Comment (achoum) [reply examples/structured_data/classification_with_tfdf.py:93]: I understood that.

However, do you think that `This attribute should *not* be used in the classifiers.` means that this field can still be used to do weighted training? This is effectively "using it" in the classifier.
I am not an expert in this dataset, so maybe is it that. Please confirm.

File: examples/structured_data/classification_with_tfdf.py (hunk 3 above)
Comment (achoum) [reply examples/structured_data/classification_with_tfdf.py:163]: I suspect that this is because those examples are training Neural networks (or another gradient descent based algorithm with mini-batching).

File: examples/structured_data/classification_with_tfdf.py (hunk 1 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:151]: Done

File: examples/structured_data/classification_with_tfdf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:163]: Done

File: examples/structured_data/classification_with_tfdf.py (hunk 5 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:172]: Done

Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:172]: Done

File: examples/structured_data/classification_with_tfdf.py (hunk 7 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py:405]: OK I added a ReLU non-linearity to the encoder, and called the it NN model, rather than linear model

**Review by achoum — APPROVED** [review]

File: examples/structured_data/classification_with_tfdf.py [hunk 8]
```diff
@@ -126,43 +115,66 @@
...
+NUM_TREES = 250  # Maximum number of decision trees. The effective number of trained trees can be smaller if early stopping is enabled.
+MIN_EXAMPLES = 6  # Minimum number of examples in a node.
+MAX_DEPTH = (
+    5  # Maximum depth of the tree. max_depth=1 means that all trees will be roots.
+)
+SUBSAMPLE = 0.65  # Ratio of the dataset (sampling without replacement) used to train individual trees for the random sampling method.
+SAMPLING_METHOD = (
+    "RANDOM"  # Control the sampling of the datasets used to train individual trees.
+)
+VALIDATION_RATIO = 0.1  # Ratio of the training dataset used to monitor the training. Require to be >0 if early stopping is enabled.
```
Comment (fchollet) [inline examples/structured_data/classification_with_tfdf.py:160]: In this section, move the comment descriptions to be above the variable (to keep lines short). Long descriptions should be split over 2 lines

File: examples/structured_data/classification_with_tfdf.py [hunk 9]
```diff
@@ -126,43 +115,66 @@
...
-
 
-def prepare_sample(features, target, weight):
-    for feature_name in features:
-        if feature_name in CATEGORICAL_FEATURES_WITH_VOCABULARY:
-            if features[feature_name].dtype != tf.dtypes.string:
-                # Convert categorical feature values to string.
-                features[feature_name] = tf.strings.as_string(features[feature_name])
-    return features, target, weight
+The `run_experiment` method is responsible loading the train and test datasets,
```
Comment (fchollet) [inline examples/structured_data/classification_with_tfdf.py:165]: `run_experiment()` (add `()`)

**Review by fchollet — APPROVED** [review]
Thank you! LGTM. There are a couple style nits, once address you can add the generated files.

File: examples/structured_data/classification_with_tfdf.py (hunk 9 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py]: Done

File: examples/structured_data/classification_with_tfdf.py (hunk 8 above)
Comment (ksalama) [reply examples/structured_data/classification_with_tfdf.py]: Done

**Comment by ksalama** [comment]
@achoum - Thank you so much for your comments. I addressed then as follows:

1. I used `CATEGORICAL_FEATURE_NAMES` instead of  `CATEGORICAL_FEATURES_WITH_VOCABULARY`
2. Create the feature semantics based on `NUMERICAL_FEATURE_NAMES` and `CATEGORICAL_FEATURES_WITH_VOCABULARY`, rather than the model input data types.
3. Deferred computing the vocabulary of the categorical features to the `create_embedding_encoder()` method.
4. Fixed the casting to float32 or float64 as we discussed
5. Updated the `BinaryTargetEncoding` to pre-compute all the required stats in the `adapt()` method, while the `call()` method would only do the lookup of the stats
6. Allow the user to provide `vocabulary_size` to avoid calling `tf.unique()`. However, if it is not passed, it's computed during the execution of the `adapt()` method. Also added a comment to recommend passing the `vocabulary_size`.
7. I kept the `prepare_dataframe()` method as it seems that ` tfdf.keras.pd_dataframe_to_tf_dataset` doesn't do all the "magic" of converting the target column to integer etc. Am I missing something?
8. Also, in the last model, when I perform activation directly on the feature, then linear projection, the results are worse than using `Dense(units, activation)(encoded_features)`. So I kept it as it was.
9. I haven't use `tf.data.experimental.group_by_reducer`, but I am planning to investigate it (in a subsequent PR).
10. Also, didn't use a dictionary for the parameters of the model as it, for some reasons, makes the example more than 300 lines, which is not allowed for the Keras examples!

Let me know if you have any further comments.

**Comment by achoum** [comment]
All good :) Thanks for the work. This is awesome.

**Review by fchollet — APPROVED** [review]
Thanks a lot to both of you! 👍
